game-core/src/main/java/games/strategy/engine/data/properties/MapProperty.java

```diff
@@ -84,36 +91,38 @@ public JComponent getViewComponent() {
 
   @Override
   public boolean validate(final Object value) {
-    if (value instanceof Map) {
-      try {
-        @SuppressWarnings("unchecked")
-        final Map<String, V> test = (Map<String, V>) value;
-        if (map != null && !map.isEmpty() && !test.isEmpty()) {
-          String key = null;
-          V val = null;
-          for (final Map.Entry<String, V> entry : map.entrySet()) {
-            if (entry.getValue() != null && entry.getKey() != null) {
-              key = entry.getKey();
-              val = entry.getValue();
-              break;
-            }
-          }
-          if (key != null) {
-            for (final Map.Entry<String, V> entry : test.entrySet()) {
-              if (entry.getKey() != null && entry.getValue() != null
-                  && (!entry.getKey().getClass().isAssignableFrom(key.getClass())
-                      || !entry.getValue().getClass().isAssignableFrom(val.getClass()))) {
-                return false;
-              }
-            }
-          }
-        }
-        resetProperties(test, new ArrayList<>(), getName(), getDescription());
-      } catch (final Exception e) {
-        return false;
-      }
-      return true;
+    if (!(value instanceof Map)) {
+      return false;
+    }
+
+    final Map<?, ?> otherMap = (Map<?, ?>) value;
+    if (!areAllElementsNullOrInstanceOf(otherMap.keySet(), String.class)
+        || !areAllElementsNullOrInstanceOf(otherMap.values(), getMapValueType())) {
+      return false;
     }
-    return false;
+
+    // verify setting new values will not trigger an error
+    try {
+      @SuppressWarnings("unchecked")
+      final Map<String, ?> typedOtherMap = (Map<String, ?>) otherMap;
+      resetProperties(typedOtherMap, new ArrayList<>(), getName(), getDescription());
+    } catch (final IllegalArgumentException e) {
+      return false;
+    }
+
+    return true;
+  }
+
+  private static boolean areAllElementsNullOrInstanceOf(final Collection<?> collection, final Class<?> type) {
+    return collection.stream()
+        .filter(not(Objects::isNull))
+        .allMatch(type::isInstance);
+  }
+
+  private Class<?> getMapValueType() {
+    return map.values().stream()
```

Comment:
I'm not checking for `null` values in this particular stream because it appears to be impossible for a `null` map value to get into this class: `resetProperties()` will throw an NPE if any map value is `null` (I think the original intention was for it to throw `IllegalArgumentException`, but an NPE is raised while constructing the IAE message).

Reply:
Fixed that in b7d40edea.